Sources/SwifterSwift/SwiftStdlib/SequenceExtensions.swift

```diff
@@ -185,17 +185,26 @@ public extension Sequence {
         }
         return (Array(matching), Array(nonMatching))
     }
+    
+    /// SwifterSwift: Return a map grouped on a keypath value.
+    ///
+    /// - Parameter by: Key path to group by.  Must be Hashable.
+    /// - Returns: Dictionary of arrays using the keypath value as the key.
+    func grouped<T: Hashable>(by keyPath: KeyPath<Element, T>) -> [T: [Element]] {
```

Comment:
@mmdock I think here we could instead of having `Sequence`, maintain the pattern of initialization and have an dictionary extension like `init<S: Sequence, T: Hashable>(grouping sequence: S, by keyPath: KeyPath<Element, T>)`... 
What you think?

Reply:
@LucianoPAlmeida I think both are valid independent extensions. 

I think `Dictionary(grouping: someSequence, by:  \String.count)` can be not as nice to read as `someSequence.grouped(by: \String.count)`, but that is my preferred way of writing swift code, which I am sure many would disagree with, so i will defer to your judgement on this.

As a note: I just tried (briefly, over my lunch break) 
`    init<S: Sequence, T: Hashable>(grouping sequence: S, by keyPath: KeyPath<Element, T>){
        self.init(grouping: sequence, by: { $0[keyPath: keyPath] })
    }`

but am receiving a `Type of expression is ambiguous without more context` on the `by: { $0[keyPath: keyPath] }`  segment.  I am not sure why this would work in the Sequence extension but not in this init 🤔.  Will think about this more when i have time this evening.

Reply:
> @LucianoPAlmeida I think both are valid independent extensions.
> 
> I think `Dictionary(grouping: someSequence, by: \String.count)` can be not as nice to read as `someSequence.grouped(by: \String.count)`, but that is my preferred way of writing swift code, which I am sure many would disagree with, so I will defer to your judgement on this.
> 

yeah, I do prefer this too ... but I think the dictionary one is preferable because is closer to the native one(the one used on the implementation). I mean if I'm a user who knows the swift stdlib has a dictionary init to do this I'll probably look for something key path based that follows the same structure. To have a native way to group the value as a Dictionary init and an extension that do the same but keyPath based as a Sequence extension seems a little inconsistent to me. So my vote is for the dictionary one 
let's see what the others think cc @omaralbeik @guykogus @chulbert 

> As a note: I just tried (briefly, over my lunch break)
> ` init<S: Sequence, T: Hashable>(grouping sequence: S, by keyPath: KeyPath<Element, T>){ self.init(grouping: sequence, by: { $0[keyPath: keyPath] }) }`
> 
> but am receiving a `Type of expression is ambiguous without more context` on the `by: { $0[keyPath: keyPath] }` segment. I am not sure why this would work in the Sequence extension but not in this init 🤔. Will think about this more when i have time this evening. If you have any obvious things to mention i can 🤦‍♂ over later, please comment it

Ohh sorry, I just typed it as a sketch... didn't know if it would be compilable...
You would have some constraints on generic types and T is actually not needed
```swift
  init<S: Sequence>(grouping sequence: S,
          by keyPath: KeyPath<S.Element, Key>) where Value == [S.Element] {
    self.init(grouping: sequence, by: { $0[keyPath: keyPath] }) 
  }
```

Reply:
@LucianoPAlmeida it took me way longer than I care to admit to understand what you were doing with that init.  I, for some reason, did not know `where` clauses can be used as part of an `init`.  I feel that that should be basic knowledge, but nope! Not me.

Implemented that and duplicated the tests.   Left in the sequence extension until we know which we prefer (or want to keep both?)

Reply:
Sure, let's see what the others think :) 
cc @omaralbeik @guykogus @chulbert

Reply:
I like them both, actually. If I had to choose I'd go for the `Dictionary` extension, which as @LucianoPAlmeida says is closer to the native initialiser.

Reply:
@guykogus @LucianoPAlmeida how about we keep both, so developers can use whichever fits their development style?  Rather than deciding for anyone wanting to use this functionality in the future.

Reply:
I'm not sure, about that ... because 
* We are implementing the same functionality twice
* Will be more work to maintain two than one, and we have to make sure the behavior is consistent
* I prefer to have only one and let the user make a wrapper in his project if it likes other style better. 

Those are the reasons I don't like the idea, but the first one is the deal breaker :) 
But ... let's see cc @omaralbeik @guykogus

Reply:
Agreed, let's not have this twice.